### PR TITLE
Add dim-by-dim subcell reconstruction

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

Dim-by-dim reconstruction is faster than all dims at once, but I think sacrifices strict conservation (I'd need to think about it more). However, it does have the added feature that constant flow is preserved along axes, which isn't the case in all dims at once reconstruction.

**Note**: The projection commit is in another PR waiting on review.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/DgSubcell/ReconstructionMethod.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  if (recons_method ==
+      get_output(evolution::dg::subcell::fd::ReconstructionMethod::DimByDim)) {
+    return type::DimByDim;
+  } else if (recons_method ==
+             get_output(evolution::dg::subcell::fd::ReconstructionMethod::
+                            AllDimsAtOnce)) {
+    return type::AllDimsAtOnce;
+  } else {
+    PARSE_ERROR(options.context(),
+                "ReconstructionMethod must be 'DimByDim', or 'AllDimsAtOnce'");
```
Comment (wthrowe) [inline src/Evolution/DgSubcell/ReconstructionMethod.cpp]: Either hardcode the valid values both here and above or use get_output in both places.

**Review by wthrowe — CHANGES_REQUESTED** [review]
You have a lot of prose here about conservation, but I don't understand it.  Isn't preserving integrals along each dimension individually a stronger condition than just preserving the multidimensional integral?

**Comment by nilsdeppe** [comment]
rebased, squashed in the `get_output` change, and changed the commit message. I agree it's a stronger constraint :)

File: src/Evolution/DgSubcell/ReconstructionMethod.hpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace evolution::dg::subcell::fd {
+/*!
+ * \brief The reconstruction method to use
+ */
+enum class ReconstructionMethod {
+  /// Dimension-by-dimension reconstruction assuming a tensor-product basis
+  DimByDim,
+  /// Reconstruct all dimensions at once requiring the integral over the
+  /// subcells and over the DG element to be equal.
```
Comment (wthrowe) [inline src/Evolution/DgSubcell/ReconstructionMethod.hpp]: Remove everything starting from "requiring".  It's not a distinguishing property of this case.

Also make the corresponding change to the commit message for the next commit.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Evolution/DgSubcell/ReconstructionMethod.hpp (hunk 2 above)
Comment (nilsdeppe) [reply src/Evolution/DgSubcell/ReconstructionMethod.hpp]: Done and rebased! Thanks :)

**Review by wthrowe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**